test-strategy.md

```diff
@@ -0,0 +1,108 @@
+# Test Strategy
+
+Very brief, mildly aspirational test strategy document. This isn't where we are but it is where we want to get to.
+
+This does not detail how to setup an environment or how to run the tests locally nor does it contain any best practices that we try to follow when writing tests, that information exists in an as yet unwritten contributing guide (ye, i know).
+
+## Objectives
+
+The purposes of all testing activities on Gradio fit one of the following objectives:
+
+1. Ensure that the Gradio library functions as we expect it to.
+2. Enable the maintenance team to quickly identify both the presence and source of defects.
+3. Prevent regressions, i.e. if we fix something it should stay fixed.
+4. Improve the quality of the codebase in order to ease maintenance efforts.
+5. Reduce the amount of manual testing required.
+
+## Scope
+
+Testing is always a tradeoff. We can't cover everything unless we want to spend all of our time writing and running tests. We should focus on a few keys areas.
+
+We should not focus on code coverage but on test coverage following the below criteria:
+
+- The documented Gradio API (that's the bit that user interact with via python) should be tested thoroughly. (1)
+- Additional gradio components that are both publicly available and used internally (such as the Python and JS client libraries) should be tested thoroughly. (1)
```

Comment:
Nit: usage of "components" can be a little unclear as it could refer to the gradio input/output components. You could use e.g. "elements" instead

Reply:
I've changed this to `elements` slightly weird but works. We can update if we think of a better word.